src/subtype.c

```diff
@@ -793,20 +793,20 @@ static int subtype_unionall(jl_value_t *t, jl_unionall_t *u, jl_stenv_t *e, int8
                 val = (jl_value_t*)u->var;
             else
                 val = (jl_value_t*)jl_new_typevar(u->var->name, vb.lb, vb.ub);
-            jl_value_t *oldval = e->envout[e->envidx];
-            // if we try to assign different variable values (due to checking
-            // multiple union members), consider the value unknown.
-            if (oldval && !jl_egal(oldval, val))
-                e->envout[e->envidx] = (jl_value_t*)u->var;
-            else
-                e->envout[e->envidx] = fix_inferred_var_bound(u->var, val);
+            if (ans) {
+                jl_value_t *oldval = e->envout[e->envidx];
+                // if we try to assign different variable values (due to checking
+                // multiple union members), consider the value unknown.
+                if (ans && oldval && !jl_egal(oldval, val))
+                    e->envout[e->envidx] = (jl_value_t*)u->var;
+                else
+                    e->envout[e->envidx] = fix_inferred_var_bound(u->var, val);
```

Comment:
I suspect #41096 might be here, since this often just returns `val`, but if `oldval` is assigned, then `val` would be wrong here (it should return `u->var` instead, or any new `TypeVar` with some `typejoin` of `val` and `oldval`)

Reply:
The behavior here is a little bit different,.
#43064 always puts the last `Union` (`union.b.b.b.b.b ...`) into `envout`
But #41096 always puts the first `Union` (`union.a`) into `envout`.
If `intersect_all` and `forall_exists_subtype` have the same traversal order, then looks like #41096 is caused by some missing `type_join`?

Reply:
#46350 has been opened for #41096.